README.md
# MSFS 2020 GPS Link

[![GitHub release][latest_release_badge]][releases_url]
[![Github all releases][downloads_badge]][releases_url]
[![CI][ci_badge]][ci]

Transmit GPS data from Microsoft Flight Simulator 2020 to navigation apps.\
Tested with [SkyDemon][sky_demon_url] and [Garmin Pilot][garmin_pilot_url].

## [Quick Download][latest_release]

## Tested on

| App                    | Platform | UDP     | COM     |
| ---------------------- | -------- | ------- | ------- |
| Garmin Pilot           | Android  | &check; | -       |
| Jeppesen FliteDeck Pro | PC       | -       | &check; |
| SkyDemon               | Android  | &check; | -       |
| SkyDemon               | PC       | &check; | &check; |
| ForeFlight Mobile      | iOS      | &check; | -       |
| ? \*                   | ?        | ?       | ?       |


\* Feedback from the ones with the ability to test will be greatly appreciated.

## Usage

- download the latest version from the [Releases][releases_url] page or by using the link above
- unzip & install
- with MSFS 2020 running, open `MSFS 2020 GPS Link` and press "Connect"

![MSFS 2020 GPS Link Usage][usage]

### SkyDemon

With `MSFS 2020 GPS Link` open and connected on your PC, open SkyDemon on your navigation device and press on `Fly` -> `Use X-Plane`.

![SkyDemon Usage][usage_skydemon]

### Garmin Pilot

With `MSFS 2020 GPS Link` open and connected on your PC, open Garmin Pilot on your navigation device and press on `Settings` -> `GPS`.

Scroll to the bottom and you should see `SIMULATOR STATUS` showing an `amber`status light. Press on `Connect`.

![Garmin Pilot Usage][usage_garmin_pilot]

## Contributing

Contributions are welcome and encouraged! Please read [CONTRIBUTING.md][contributing_url].

[latest_release]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/releases/download/v0.3.2/msfs-2020-gps-link-v0.3.2.zip
[latest_release_badge]: https://img.shields.io/github/release/mihai-dinculescu/msfs-2020-gps-link.svg
[downloads_badge]: https://img.shields.io/github/downloads/mihai-dinculescu/msfs-2020-gps-link/total.svg
[ci_badge]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/workflows/CI/badge.svg?branch=main
[ci]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/actions
[sky_demon_url]: https://www.skydemon.aero
[garmin_pilot_url]: https://buy.garmin.com/en-US/US/p/115856
[releases_url]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/releases
[usage]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/blob/main/assets/usage.PNG
[usage_skydemon]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/blob/main/assets/usage-skydemon.PNG
[usage_garmin_pilot]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/blob/main/assets/usage-garmin-pilot.PNG
[contributing_url]: /CONTRIBUTING.md


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this
file. This change log follows the conventions of
[keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

## [v0.3.2] - 2023-03-22

## Changed

- Updated to `simconnect-sdk` v0.2.2.
- Updated to MSFS SDK v0.20.5.0.

## Fixed

- Fixed an issue that was causing the heading to be inaccurate with easterly magnetic variations.

## [v0.3.1] - 2022-11-14

### Fixed

- GPS Ground Speed is now correctly sent in knots instead of meters per second over COM.

## [v0.3.0] - 2022-11-07

### Added

- The option to broadcast to a virtual COM port, emulating an RS232 GPS device, has been added.

### Changed

- When a new release is available, the notification now includes its version number.

## [v0.3.0-alpha.2] - 2022-11-06

### Changed

- The **Baud rate** field is now a dropdown with a fixed set of options.

### Fixed

- The values for **Course Over Ground** and **Magnetic Variation** sent over COM should now be correct.

## [v0.3.0-alpha.1] - 2022-11-06

### Added

- The option to broadcast to a virtual COM port, emulating an RS232 GPS device, has been added.

### Changed

- When a new release is available, the notification now includes its version number.

## [v0.2.5] - 2022-10-29

### Changed

- The "Connected" status will now be reported much quicker (up to 6x times faster).
- Error handling and tracing have been improved.

### Fixed

- Fixed an issue that was causing the UI to occasionally think it's still connected after MSFS2020 has been closed.
- Fixed an issue that was preventing the SimConnect Client from properly closing on user "Disconnect".
- Fixed an issue that was causing SimConnect Client to unnecessarily connect a second time on user "Connect".

## [v0.2.4] - 2022-08-26

### Changed

- All dependencies are updated to the latest version.

## [v0.2.3] - 2021-05-31

### Changed

- The GPS Track is now calculated by subtracting MAGVAR from GPS GROUND MAGNETIC TRACK

## [v0.2.2] - 2021-05-05

### Changed

- The GPS track is now based on 'GPS GROUND MAGNETIC TRACK' instead of 'GPS GROUND TRUE TRACK'

## [v0.2.1] - 2021-05-01

### Changed

- The GPS track is now based on 'GPS GROUND TRUE TRACK' instead of 'PLANE HEADING DEGREES TRUE'

## [v0.2.0] - 2021-02-21

### Added

- SkyDemon documentation
- Garmin Pilot documentation

### Changed

- Rename "Broadcast subnet mask" to "Broadcast address" and add an explanation for how it works

### Fixed

- Fix the issue that is causing the broadcast to happen on the "255.255.255.255" netmask, irrespective of what is configured in the UI
- Fix the text input mask of "Broadcast address"
- Fix the issue that is causing the check for a new version to happen more often than needed

## [v0.1.0] - 2020-12-26

### Initial Release of MSFS 2020 GPS Link

[unreleased]: https://github.com/mihai-dinculescu/msfs-2020-gps-link
[v0.3.2]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.3.2
[v0.3.1]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.3.1
[v0.3.0]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.3.0
[v0.3.0-alpha.2]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.3.0-alpha.2
[v0.3.0-alpha.1]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.3.0-alpha.1
[v0.2.5]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.2.5
[v0.2.4]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.2.4
[v0.2.3]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.2.3
[v0.2.2]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.2.2
[v0.2.1]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.2.1
[v0.2.0]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.2.0
[v0.1.0]: https://github.com/mihai-dinculescu/msfs-2020-gps-link/tree/v0.1.0


src/broadcaster/com.rs
use std::{fmt, io, time::SystemTime};

use chrono::{DateTime, Utc};
use serialport::SerialPort;
use tracing::{error, instrument, trace, warn};

use crate::system::simconnect_objects::GpsData;

use super::{BroadcasterExt, ComConfig};

const NMEA_MID_GSA_MESSAGE: &str =
    "$GPGSA,A,3,01,02,03,04,05,06,07,08,09,10,11,12,1.0,1.0,1.0*30\r\n";
const NMEA_MID_GSA_INTERVAL_S: u64 = 1;

pub struct Com {
    port: Box<dyn SerialPort>,
    last_mid_gsa: Option<SystemTime>,
}

impl fmt::Debug for Com {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Com")
            .field("port", &"...")
            .field("last_mid_gsa", &self.last_mid_gsa)
            .finish()
    }
}

impl Com {
    #[instrument(name = "Com::new")]
    pub fn new(config: ComConfig) -> Result<Box<dyn BroadcasterExt>, io::Error> {
        let port = serialport::new(config.port, config.baud_rate)
            .open()
            .map_err(|e| {
                error!(error = ?e, "failed to open COM port");
                e
            })?;

        Ok(Box::new(Com {
            port,
            last_mid_gsa: None,
        }))
    }
}

impl BroadcasterExt for Com {
    #[instrument(name = "Com::send", skip(self, data))]
    fn send(&mut self, data: GpsData) -> Result<(), io::Error> {
        if self.should_send_mid_gsa() {
            self.write(NMEA_MID_GSA_MESSAGE)?;
            self.last_mid_gsa.replace(SystemTime::now());
        }

        let date = Utc::now();

        let message = Self::convert_gps_data_to_nmea_mid_gga(&date, &data);
        self.write(&message)?;

        let message = Self::convert_gps_data_to_nmea_mid_rmc(&date, &data);
        self.write(&message)?;

        trace!("Successfully sent broadcast over COM");

        Ok(())
    }
}

impl Com {
    fn should_send_mid_gsa(&self) -> bool {
        match self.last_mid_gsa {
            Some(last_mid_gsa) => {
                if let Ok(elapsed) = last_mid_gsa.elapsed() {
                    if elapsed.as_secs() >= NMEA_MID_GSA_INTERVAL_S {
                        return true;
                    }
                }
            }
            None => {
                return true;
            }
        }

        false
    }

    fn write(&mut self, data: &str) -> Result<(), io::Error> {
        match self.port.write(data.as_bytes()) {
            Ok(_) => (),
            Err(ref e) if e.kind() == io::ErrorKind::TimedOut => {
                warn!(error = ?e, "failed to write to COM (timeout)")
            }
            Err(e) => {
                error!(error = ?e, "failed to write to COM");
                return Err(e);
            }
        };

        Ok(())
    }

    fn convert_gps_data_to_nmea_mid_gga(date: &DateTime<Utc>, data: &GpsData) -> String {
        let (lat_deg, lat_min, lat_dir) = data.lat_as_degrees_minutes_dir();
        let (lon_deg, lon_min, lon_dir) = data.lon_as_degrees_minutes_dir();

        let message = format!(
            "$GPGGA,{},{:0>2}{:0>7.4},{},{:0>3}{:0>7.4},{},1,12,1.0,{:.1},M,0.0,M,,",
            date.format("%H%M%S%.3f"),
            lat_deg,
            lat_min,
            lat_dir,
            lon_deg,
            lon_min,
            lon_dir,
            data.pressure_altitude
        );

        let checksum = message.chars().skip(1).fold(0u8, |acc, c| acc ^ c as u8);

        format!("{message}*{checksum:X}\r\n")
    }

    fn convert_gps_data_to_nmea_mid_rmc(date: &DateTime<Utc>, data: &GpsData) -> String {
        let (lat_deg, lat_min, lat_dir) = data.lat_as_degrees_minutes_dir();
        let (lon_deg, lon_min, lon_dir) = data.lon_as_degrees_minutes_dir();
        let (magnetic_variation, magnetic_variation_dir) = data.gps_magnetic_variation_as_abs_dir();

        let message = format!(
            "$GPRMC,{},A,{:0>2}{:0>7.4},{},{:0>3}{:0>7.4},{},{:.2},{:.2},{},{:.1},{},S",
            date.format("%H%M%S%.3f"),
            lat_deg,
            lat_min,
            lat_dir,
            lon_deg,
            lon_min,
            lon_dir,
            data.gps_ground_speed_in_knots(),
            data.gps_ground_true_track,
            date.format("%d%m%y"),
            magnetic_variation,
            magnetic_variation_dir,
        );

        let checksum = message.chars().skip(1).fold(0u8, |acc, c| acc ^ c as u8);

        format!("{message}*{checksum:X}\r\n")
    }
}

#[cfg(test)]
mod tests {
    use chrono::{NaiveDate, Utc};

    use super::Com;

    #[test]
    fn test_convert_gps_data_to_nmea_mid_gga() {
        let date = NaiveDate::from_ymd_opt(2022, 10, 30)
            .and_then(|d| d.and_hms_milli_opt(21, 10, 30, 750))
            .unwrap()
            .and_local_timezone(Utc)
            .unwrap();

        let data = crate::system::simconnect_objects::GpsData {
            lat: 51.509865,
            lon: -0.118092,
            alt: 0.0,
            pressure_altitude: 3.0,
            gps_ground_true_track: 310.55,
            gps_magnetic_variation: 5.0,
            gps_ground_speed: 100.50,
        };

        let result = Com::convert_gps_data_to_nmea_mid_gga(&date, &data);

        assert_eq!(
            result,
            "$GPGGA,211030.750,5130.5919,N,00007.0855,W,1,12,1.0,3.0,M,0.0,M,,*70\r\n"
        );
    }

    #[test]
    fn test_convert_gps_data_to_nmea_mid_gga_2() {
        let date = NaiveDate::from_ymd_opt(2022, 1, 3)
            .and_then(|d| d.and_hms_milli_opt(2, 1, 3, 75))
            .unwrap()
            .and_local_timezone(Utc)
            .unwrap();

        let data = crate::system::simconnect_objects::GpsData {
            lat: 0.00040752447554520855,
            lon: 0.01397450300629543,
            alt: 0.0,
            pressure_altitude: 0.9642891859251844,
            gps_ground_true_track: 92.71680515837362,
            gps_magnetic_variation: -4.384223296150313,
            gps_ground_speed: 0.0,
        };

        let result = Com::convert_gps_data_to_nmea_mid_gga(&date, &data);

        assert_eq!(
            result,
            "$GPGGA,020103.075,0000.0245,N,00000.8385,E,1,12,1.0,1.0,M,0.0,M,,*68\r\n"
        );
    }

    #[test]
    fn test_convert_gps_data_to_nmea_mid_rmc() {
        let date = NaiveDate::from_ymd_opt(2022, 10, 30)
            .and_then(|d| d.and_hms_milli_opt(21, 10, 30, 750))
            .unwrap()
            .and_local_timezone(Utc)
            .unwrap();

        let data = crate::system::simconnect_objects::GpsData {
            lat: 51.509865,
            lon: -0.118092,
            alt: 0.0,
            pressure_altitude: 3.0,
            gps_ground_true_track: 310.55,
            gps_magnetic_variation: 5.0,
            gps_ground_speed: 100.50,
        };

        let result = Com::convert_gps_data_to_nmea_mid_rmc(&date, &data);

        assert_eq!(
            result,
            "$GPRMC,211030.750,A,5130.5919,N,00007.0855,W,195.36,310.55,301022,5.0,E,S*7\r\n"
        );
    }

    #[test]
    fn test_convert_gps_data_to_nmea_mid_rmc_2() {
        let date = NaiveDate::from_ymd_opt(2022, 1, 3)
            .and_then(|d| d.and_hms_milli_opt(2, 1, 3, 75))
            .unwrap()
            .and_local_timezone(Utc)
            .unwrap();

        let data = crate::system::simconnect_objects::GpsData {
            lat: 0.00040752447554520855,
            lon: 0.01397450300629543,
            alt: 0.0,
            pressure_altitude: 0.9642891859251844,
            gps_ground_true_track: 92.71680515837362,
            gps_magnetic_variation: -4.384223296150313,
            gps_ground_speed: 0.0,
        };

        let result = Com::convert_gps_data_to_nmea_mid_rmc(&date, &data);

        assert_eq!(
            result,
            "$GPRMC,020103.075,A,0000.0245,N,00000.8385,E,0.00,92.72,030122,4.4,W,S*3E\r\n"
        );
    }
}
